tool/planet/membership.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	serf "github.com/hashicorp/serf/client"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// startSerfReconciler creates a control loop that periodically attempts to
+// reconcile serf cluster.
+func startSerfReconciler(ctx context.Context, kubeconfig *rest.Config, serfConfig *serf.Config) {
+	const reconcileInterval = time.Minute * 10
+	ticker := time.NewTicker(reconcileInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := reconcileSerfCluster(kubeconfig, serfConfig); err != nil {
+				log.WithError(err).Warn("Failed to reconcile serf cluster.")
+			}
+		case <-ctx.Done():
+			log.Debug("Stopping serf reconciler")
+			return
+		}
+	}
+}
+
+// reconcileSerfCluster reconciles ther serf cluster if any members are missing
+// from the serf cluster.
+func reconcileSerfCluster(kubeconfig *rest.Config, serfConfig *serf.Config) error {
+	k8sClient, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		return trace.Wrap(err, "failed to create kubernetes clientset")
+	}
+
+	serfClient, err := serf.ClientFromConfig(serfConfig)
+	if err != nil {
+		return trace.Wrap(err, "failed to create serf client")
+	}
+	defer serfClient.Close()
+
+	return reconcileSerf(k8sClient, serfClient)
+}
+
+// reconcileSerf reconciles the serf cluster if any members are missing from the
+// cluster.
+func reconcileSerf(k8sClient kubernetes.Interface, serfClient serfClient) error {
+	serfPeers, err := getSerfPeers(serfClient)
+	if err != nil {
+		return trace.Wrap(err, "failed to get serf peers")
+	}
+
+	k8sPeers, err := getK8sPeers(k8sClient)
+	if err != nil {
+		return trace.Wrap(err, "failed to get k8s peers")
+	}
+
+	if !shouldReconcileSerf(k8sPeers, serfPeers) {
+		return nil
+	}
+
+	if _, err := serfClient.Join(k8sPeers, false); err != nil {
```

Comment:
I'm just curious, why are we joining to all peers, instead of having `shouldReconcileSerf` return a list of missing peers and join just those?

Then you wouldn't also need the `getSerfPeers` call below as you'd just be able to log the serf cluster and the missing peers.

Reply:
Good point. No need to join all peers.